File: fintechplatform-ui/src/main/java/com/fintechplatform/ui/payout/ui/PayOutUI.java
package com.fintechplatform.ui.payout.ui;

import android.content.Context;
import android.content.Intent;

import com.android.volley.toolbox.Volley;
import com.fintechplatform.api.account.balance.api.BalanceAPIModule;
import com.fintechplatform.api.cashout.api.CashOutAPIModule;
import com.fintechplatform.api.iban.api.IbanAPIModule;
import com.fintechplatform.api.net.NetModule;
import com.fintechplatform.ui.iban.ui.IbanUI;
import com.fintechplatform.ui.models.DataAccount;

public class PayOutUI {

    private String hostname;
    private IbanUI ibanUI;
    private DataAccount configuration;

    protected static PayOutUI instance;

    protected PayOutUI() {
    }

    public PayOutUI(String hostname, DataAccount configuration, IbanUI ibanUI) {
        this.ibanUI = ibanUI;
        this.configuration = configuration;
        this.hostname = hostname;
    }

    protected PayOutViewComponent buildPayOutViewComponent(Context context, PayOutContract.View view) {
        return DaggerPayOutViewComponent.builder()
                .payOutPresenterModule(new PayOutPresenterModule(view, instance.configuration))
                .cashOutAPIModule(new CashOutAPIModule(instance.hostname))
                .netModule(new NetModule(new NetData(Volley.newRequestQueue(context), hostname)))
                .balanceAPIModule(new BalanceAPIModule(instance.hostname))
                .ibanAPIModule(new IbanAPIModule(instance.hostname))
                .ibanUIModule(new IbanUIModule(ibanUI))
                .build();
    }

    public static PayOutViewComponent createPayOutViewComponent(Context context, PayOutContract.View view) {
        return instance.buildPayOutViewComponent(context, view);
    }

    public void start(Context context) {
        instance = this;
        Intent intent = new Intent(context, PayOutActivity.class);
        context.startActivity(intent);
    }

}

File: fintechplatform-ui/src/main/java/com/fintechplatform/ui/payout/PayOutUI.kt
package com.fintechplatform.ui.payout

import android.content.Context
import android.content.Intent
import android.os.Bundle
import com.android.volley.toolbox.Volley
import com.fintechplatform.api.account.balance.api.BalanceAPIModule
import com.fintechplatform.api.cashout.api.CashOutAPIModule
import com.fintechplatform.api.iban.api.IbanAPIModule
import com.fintechplatform.api.net.NetModule
import com.fintechplatform.ui.models.DataAccount
import com.fintechplatform.ui.payout.di.DaggerPayOutViewComponent
import com.fintechplatform.ui.payout.di.PayOutPresenterModule
import com.fintechplatform.ui.payout.di.PayOutViewComponent


class PayOutUI( val hostName: String, 
                val configuration: DataAccount) {

    fun startActivity(context: Context) {
        val intent = Intent(context, PayOutActivity::class.java)
        val args = Bundle()
        args.putString("hostname", hostName)
        args.putParcelable("dataAccount", configuration)
        intent.putExtras(args)
        context.startActivity(intent)
    }

    fun createFragment() : PayOutFragment {
        return PayOutFragment.newInstance(hostName, configuration)
    }

    object Builder {
        /**
         * Resolve with default decendencies for pay out in view component
         */

        fun buildPayOutViewComponent(context: Context, view: PayOutContract.View, hostName: String, dataAccount: DataAccount): PayOutViewComponent {
            return DaggerPayOutViewComponent.builder()
                    .payOutPresenterModule(PayOutPresenterModule(view, dataAccount))
                    .cashOutAPIModule(CashOutAPIModule(hostName))
                    .netModule(NetModule(Volley.newRequestQueue(context), hostName))
                    .balanceAPIModule(BalanceAPIModule(hostName))
                    .ibanAPIModule(IbanAPIModule(hostName))
                    .build()
        }
    }

}